---
title: "KSPZV1 DGE Δ (postvax/baseline) limma voom - revision"
author: "Tuan M. Tran"
date: "05/16/2022"
output: html_document
---

### Objective

Perform differential gene expression using limma voom to assess differences between protected and not protected infants in the KSPZV1 malaria vaccine trial.
This analysis specifically evaluates differences in post-vax with adjustment for baseline.

This analysis also includes adjustments for:

1. Batch
2. Sex
3. Study site
4. log10(CSP Ab + 1)
5. number of malaria infections during vaccination period

### Load required packages


```{r libraries, message=FALSE, warning=FALSE}
library(knitr)
library(tidyverse)
library(limma)
library(edgeR)
library(googledrive)
library(tidyverse)
library(fgsea)
library(data.table)
library(EDASeq)
library(gtools)
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Load ExpressionSet

```{r load SeqExpressionSet 2, message=FALSE, warning=FALSE}
#from google drive
#local path
#/Volumes/GoogleDrive/My Drive/Tran Lab Shared/Projects/Doris Duke PfSPZ Kenya/Tuan PfSPZ/KenyaPfSPZ/Final Data Files/KSPZV1 SeqExpressionSet Baseline Postvax Reduced Phenodat 35716x487.rds

temp <- tempfile(fileext = ".rds")
dl <- drive_download(
  as_id("17P7RwDaPCwDXcrD82N4fYd3vGOcueFEH"), path = temp, overwrite = TRUE)
x <- readRDS(file = dl$local_path)
x$treat <- factor(x$treat, levels = c("Placebo", "4.5 x 10^5 PfSPZ", "9.0 x 10^5 PfSPZ", "1.8 x 10^6 PfSPZ"))
```

### Set options

```{r set options, message=FALSE, warning=FALSE}
myGroups <- unique(x$treat)
```


```{r more reduction, echo=FALSE}
xgroup <- setNames(as.list(c(1:4)), c("Placebo", "4.5 x 10^5 PfSPZ", "9.0 x 10^5 PfSPZ", "1.8 x 10^6 PfSPZ"))
for(i in names(xgroup)){
  xgroup[[i]] <- x[, which(x$treat == i)]
  xgroup[[i]] <- xgroup[[i]][,xgroup[[i]]$Timepoint == 0 | xgroup[[i]]$Timepoint == 25]
  xgroup[[i]]$treat <- droplevels(xgroup[[i]]$treat)
  print(i)
  print(table(Outcome = xgroup[[i]]$Outcome, Timepoint = xgroup[[i]]$Timepoint))
  print(table(Outcome = xgroup[[i]]$Outcome, Dosegroup = xgroup[[i]]$treat))
  print(dim(xgroup[[i]]))
}
```

### Reduce to high dose (1.8 x 10^6 PfSPZ) only

This reduces the time and enables the code to be knitted without throwing errors related to url connections. Can leave out if all treatment groups are analyzed when script is ran locally.

```{r reduce to high dose only, message=FALSE, warning=FALSE}
myList <- c("1.8 x 10^6 PfSPZ")
for (name in myList) {
  xgroup <- xgroup[name]
}
```

### Build DGEList Object

```{r make DGEList object, message=FALSE, warning=FALSE}
#Define group by dose, timepoint and outcome
ygroup <- xgroup
for(i in names(ygroup)){
  ygroup[[i]]  <- DGEList(counts=counts(xgroup[[i]]), genes=fData(xgroup[[i]]), group= factor(paste(xgroup[[i]]$treat, xgroup[[i]]$Timepoint, xgroup[[i]]$Outcome, sep = "_")), remove.zeros=T)
  ygroup[[i]]$samples$PATID <- gsub("_.*", "", rownames(ygroup[[i]]$samples))
}
```

### Filter out low expression features

```{r low expression filters, message=FALSE, warning=FALSE}
for(i in names(ygroup)){
  keep <- filterByExpr(ygroup[[i]])
  ygroup[[i]] <- ygroup[[i]][keep, , keep.lib.sizes=FALSE]
  ygroup[[i]]$samples$lib.size <- colSums(ygroup[[i]]$counts)
  ygroup[[i]] <- calcNormFactors(ygroup[[i]])	#Normalization
}
```

### Remove unpaired samples and then arrange by patient and timepoint

This is necessary for paired analyses

```{r remove unpaired samples then arrange, message=FALSE, warning=FALSE}
for(i in names(ygroup)){
  ygroup[[i]] <- ygroup[[i]][,!is.na(ygroup[[i]]$samples$group)]
  ygroup[[i]] <- ygroup[[i]][,duplicated(ygroup[[i]]$samples$PATID) | duplicated(ygroup[[i]]$samples$PATID, fromLast = TRUE)]
  ygroup[[i]] <- ygroup[[i]][,order(ygroup[[i]]$samples$PATID, ygroup[[i]]$samples$group)]
  print(i)
  print(ifelse(all(gsub('_.*', '', colnames(ygroup[[i]][,grepl("_0", colnames(ygroup[[i]]))])) ==
               gsub('_.*', '', colnames(ygroup[[i]][,grepl("_25", colnames(ygroup[[i]]))]))) &
           all(ygroup[[i]]$samples$Timepoint == rep(c(0,25),length(ygroup[[i]]$samples$Timepoint)/2)), 
         "Samples are paired. All is good.","Stop and check order of samples."))
}
```

### Remake ExpressionSet after sample and feature (gene) filtering steps.

```{r remake SeqExpressionSet after gene filtering and sample reduction, message=FALSE, warning=FALSE}
for(i in names(ygroup)){
  myintersection <- intersect(rownames(pData(xgroup[[i]])), colnames(ygroup[[i]]$counts))
  xgroup[[i]] <- newSeqExpressionSet(counts = ygroup[[i]]$counts[,myintersection],
                                     featureData = AnnotatedDataFrame(fData(xgroup[[i]])[which(rownames(fData(xgroup[[i]])) %in% rownames(ygroup[[i]]$counts)),]),
                                     phenoData = AnnotatedDataFrame(pData(xgroup[[i]])[myintersection,]))
}
```

### Design Matrix - Compare Protected vs Not Protected

Analysis between Protected and Not Protected within a Dose Group.

Set up design for limma voom. Batch, study site, CSP-specific IgG pre-immunization, and malaria during vaccination period as fixed effects. Subject as random effect using the duplicateCorrelation function.See following:

https://bioconductor.riken.jp/packages/3.9/bioc/vignettes/variancePartition/inst/doc/dream.html

Chapter 9.7 Multi-level Experiments: https://www.bioconductor.org/packages/devel/bioc/vignettes/limma/inst/doc/usersguide.pdf

```{r design matrix protected vs susceptible by group, message=FALSE, warning=FALSE}

DeltaDelta <- DeltaNotProtected <- DeltaProtected <- fit <- design <- xgroup

for(i in names(xgroup)){
  Subject <- factor(xgroup[[i]]$PATID)
  Outcome <- factor(xgroup[[i]]$Outcome, levels = c(1,0), labels = c("NotProtected","Protected"))
  Timepoint <- factor(xgroup[[i]]$Timepoint, levels = c(0,25), labels = c("baseline","postvax"))
  Batch <- factor(xgroup[[i]]$SEQBATCH, levels = c("Aug2019","Nov2019"))
  Sex <- factor(xgroup[[i]]$SEX)
  Site <- factor(xgroup[[i]]$site)
  Pf_VAX1 <- factor(xgroup[[i]]$mal.vax.1)
  MALdVaxTotal <- as.numeric(ifelse(is.na(xgroup[[i]]$mal.dvax.tot), median(xgroup[[i]]$mal.dvax.tot, na.rm = TRUE), xgroup[[i]]$mal.dvax.tot)) #Impute  missing values with median of all samples (there are no missing values in high-dose PfSPZ group)
  MALdVax <- factor(xgroup[[i]]$mal.dvax)
  CSPAb_baseline <- as.numeric(xgroup[[i]]$pfcsp_pre)
  CSPAb_baseline[is.na(CSPAb_baseline)] <- median(CSPAb_baseline, na.rm = TRUE) #Impute 2 missing values with median of all samples
  CSPAb_baseline <- log10(CSPAb_baseline+1) #log transform to make linear
  #Define protective outcome-specific PfSPZ Vaccination effects and append them to the design matrix
  Protected.Postvax <- Outcome == "Protected" & Timepoint=="postvax"
  NotProtected.Postvax <- Outcome =="NotProtected" & Timepoint=="postvax"
  design[[i]] <- model.matrix(~Batch+Sex+Site+CSPAb_baseline+MALdVaxTotal)
  design[[i]] <- cbind(design[[i]], Protected.Postvax, NotProtected.Postvax) #make design matrix according to edgeR 3.5 (p42)
  #https://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf
  rownames(design[[i]]) <- xgroup[[i]]$PATID
  print(paste0("running model fit for ", i))
  print(colnames(design[[i]])) #check colnames

  #See 18.1.9 of Limma user guide
  #Use voom() to convert the read counts to log2-cpm, with associated weights, ready for linear modeling
  
  v <- voom(ygroup[[i]], design[[i]])
  cor <- duplicateCorrelation(v, design[[i]], block = Subject)
  print(cor$consensus)
  v <- voom(ygroup[[i]], design[[i]], plot = TRUE, block = Subject, correlation = cor$consensus)
  cor <- duplicateCorrelation(v, design[[i]], block = Subject)
  print(cor$consensus)
  
  #fit
  fit <- lmFit(v, design[[i]], block = Subject, correlation = cor$consensus)
  fit <- eBayes(fit)
  DeltaProtected[[i]] <- topTable(fit,n=Inf, coef="Protected.Postvax")
  DeltaNotProtected[[i]] <-  topTable(fit,n=Inf, coef="NotProtected.Postvax")
  fit2 <- contrasts.fit(fit, contrasts = c(rep(0,(ncol(design[[i]])-2)),1,-1))
  fit2 <- eBayes(fit2)
  DeltaDelta[[i]] <- topTable(fit2,n=Inf)
}
```

```{r save data, echo=FALSE, eval=FALSE}
# save(design, fit, fit2, DeltaProtected, DeltaNotProtected, DeltaDelta, file = "/Volumes/GoogleDrive/My Drive/Manuscripts/KSPZV1 Manuscript/Med Revised Submission 2022/Revision R Objects/DeltaDelta Model highdose PfSPZ only Batch Sex Site CSPAb MALdVaxTotal adj.RData")
```


### Examine top 15 DEGs in 1.8x10^6 PfSPZ dose group

Δ Protected 

```{r DEG table placebo, echo = TRUE}
DeltaProtected$`1.8 x 10^6 PfSPZ`[1:15,c(7,9,2:5,8,10:15)] %>%
  mutate_if(is.numeric, signif,3) %>%
  knitr::kable(., row.names = FALSE)
```

Δ Not Protected

```{r DEG table low dose, echo = TRUE}
DeltaNotProtected$`1.8 x 10^6 PfSPZ`[1:15,c(7,9,2:5,8,10:15)] %>%
  mutate_if(is.numeric, signif,3) %>%
  knitr::kable(., row.names = FALSE)
```

Δ Protected vs. Δ Not Protected

```{r DEG table med dose, echo = TRUE}
DeltaDelta$`1.8 x 10^6 PfSPZ`[1:15,c(7,9,2:5,8,10:15)] %>%
  mutate_if(is.numeric, signif,3) %>%
  knitr::kable(., row.names = FALSE)
```


```{r make table s3 delta degs for highdose, echo=FALSE, eval=FALSE}
resdir <- "/Volumes/GoogleDrive/My Drive/Manuscripts/KSPZV1 Manuscript/Med Revised Submission 2022/Supplementary Tables/"

dp_bound <- DeltaProtected %>%
  bind_rows(., .id = "dosegroup") %>%
  mutate(comparison = "ΔP") %>%
  mutate(study = "KSPZV1") %>%
  mutate(timepoint = "delta (post-immunization/pre-immunization baseline")

dnp_bound <- DeltaNotProtected %>%
  bind_rows(., .id = "dosegroup") %>%
  mutate(comparison = "ΔNP") %>%
  mutate(study = "KSPZV1") %>%
  mutate(timepoint = "delta (post-immunization/pre-immunization baseline")

dd_bound <- DeltaDelta %>%
  bind_rows(., .id = "dosegroup") %>%
  mutate(comparison = "ΔP vs ΔNP") %>%
  mutate(study = "KSPZV1") %>%
  mutate(timepoint = "delta (post-immunization/pre-immunization baseline")

all_bound <- bind_rows(dp_bound, dnp_bound, dd_bound) %>%
  dplyr::select(study, dosegroup, timepoint, comparison, GeneSymbol, EnsemblID, descripton_new, everything()) %>%
  dplyr::select(-c(Description, Chromosome, Start, End, Probe.Strand, Geneid)) %>%
  dplyr::rename(ensembl_id = "EnsemblID") %>%
  dplyr::rename(description = "descripton_new") %>%
  dplyr::rename(hgnc_symbol = "GeneSymbol") %>%
  as_tibble()

all_bound %>%
  dplyr::select(study, dosegroup, timepoint, comparison, hgnc_symbol, ensembl_id, description, logFC, AveExpr, t, B, P.Value, adj.P.Val) %>%
  writexl::write_xlsx(., path = paste0(resdir, "Table S3 Delta DEG Tables for high-dose PfSPZ KSPZV1 REVISED.xlsx"))
```
### Apply GSEA

Rank genes by -log10(P.Value)*sign(logFC). Run fgsea from fgsea package using NamedGeneRankList2GseaTable helper function.

```{r apply fgsea, warning=FALSE, message=FALSE}
set.seed(23)
#restructure dataframes
degtabs <- ygroup
for(i in names(ygroup)){
    degtabs[[i]] <- setNames(as.list(c(1:3)), c("DeltaDelta", "DeltaP", "DeltaNP"))
    degtabs[[i]]$DeltaDelta <- DeltaDelta[[i]]
    degtabs[[i]]$DeltaP <- DeltaProtected[[i]]
    degtabs[[i]]$DeltaNP <- DeltaNotProtected[[i]]
}
ranks <- degtabs
for(i in names(degtabs)){
  for(j in names(degtabs[[i]])){
    ranks[[i]][[j]] <-  degtabs[[i]][[j]] %>%
      mutate(rankmetric = -log10(.$P.Value)*sign(.$logFC)) %>%
      dplyr::select(GeneSymbol,rankmetric) %>%
      na.omit() %>% 
      distinct() %>% 
      group_by(GeneSymbol) %>%
      summarize(rankmetric = mean(rankmetric)) %>%
      arrange(desc(rankmetric))
  }
}
#Run fgsea from fgsea package using NamedGeneRankList2GseaTable helper function
devtools::source_url("https://github.com/TranLab/ModuleLists/blob/main/NamedGeneRankList2GseaTable.R?raw=TRUE")
GSEAtab <- ranks
for(i in names(ranks)){
  for(j in names(ranks[[i]])){
    print(paste0("starting ", j, " in ", i))
    GSEAtab[[i]][[j]] <- NamedGeneRankList2GseaTable(rankedgenes = deframe(ranks[[i]][[j]]),
                                                     geneset = "bloodmodules",
                                                     output_directory = tempdir(),
                                                     filename_prefix = "GSEA",
                                                     sampleSize = 101,
                                                     minSize = 20,
                                                     maxSize = Inf,
                                                     scoreType = "std") %>%
      as_tibble() %>%
      arrange(desc(NES)) %>% 
      dplyr::select(module_type, pathway, ES, NES, size, leadingEdge, pval, padj) %>% 
      mutate(leadingEdge = gsub("^c\\(|\\)$", "", leadingEdge)) %>%
      mutate(leadingEdge = gsub('"', "", leadingEdge)) %>%
      arrange(padj)
    #closeAllConnections() #this line prevents using up all url connections
    }
}
```

```{r make table sx delta gsea for highdose, echo=FALSE, eval=FALSE}
allGSEA <- GSEAtab$`1.8 x 10^6 PfSPZ` %>%
  bind_rows(., .id = "comparison") %>%
  mutate(comparison = gsub("DeltaDelta", "ΔP vs ΔNP", comparison)) %>%
  mutate(comparison = gsub("Delta", "Δ", comparison)) %>%
  mutate(study = "KSPZV1") %>%
  mutate(dosegroup = "1.8 x 10^6 PfSPZ") %>%
  mutate(timepoint = "delta (post-immunization/pre-immunization baseline") %>%
  dplyr::select(study, dosegroup, timepoint, comparison, everything())

allGSEA %>%
  writexl::write_xlsx(., path = paste0(resdir, "Table SX Delta GSEA Tables for high-dose PfSPZ KSPZV1 REVISED.xlsx"))
```
### Visualize GSEA data as bubble plot

Filter based on padj < 0.20.

```{r plot bubble plot, echo = TRUE, message=FALSE, warning=FALSE}
plotDat <- GSEAtab$`1.8 x 10^6 PfSPZ` %>%
  bind_rows(., .id = "comparison") %>%
  mutate(comparison = factor(comparison, levels = c("DeltaP", "DeltaNP", "DeltaDelta"), labels = c("DeltaP", "DeltaNP", "DeltaDelta"))) %>%
  filter(padj < 0.20) %>%
  filter(module_type %in% c("highBTMs", "lowBTMs", "MonacoModules", "BloodGen3Module"))  %>%
  filter(!grepl("TBA", pathway)) %>%
  dplyr::select(comparison, module_type, pathway, leadingEdge, size, NES, padj) %>%
  mutate(neglogpadj = -log10(padj)) %>%
  mutate(pathway = gsub("gd", "γδ", pathway)) %>%
  mutate(pathway = gsub("Vd2", "Vδ2", pathway)) %>%
  mutate(pathway = gsub("Vg", "Vγ", pathway)) %>%
  mutate(pathway = gsub("HALLMARK_", "", pathway)) %>%
  mutate(pathway = gsub("_", " ", pathway)) %>%
  mutate(pathway = sub(".*?\\_", "", pathway)) %>%
  group_by(module_type) %>%
  mutate(pathway = fct_reorder(pathway, NES, .desc = TRUE)) %>%
  ungroup() %>%
  filter(!grepl("TBD", pathway)) %>%
  mutate(module_type = factor(module_type, levels = c("highBTMs", "MonacoModules", "lowBTMs", "BloodGen3Module"))) %>%
  arrange(desc(neglogpadj)) %>% 
  droplevels()

foodat <- plotDat %>%
  filter(comparison == "DeltaNP")
#plotting options
basetextsize <- 8  
myfont <- "sans"
bubble_max_size <- 6

MyPlot <- plotDat %>%
  filter(module_type %in% c("highBTMs", "MonacoModules"))  %>%
  ggplot(., aes(x = comparison, y = pathway)) +
  geom_point(aes(size=neglogpadj, fill = NES), alpha = 0.65, shape=21, stroke = 0.25) +
  scale_size_area(name = expression(-log[10]~adj.~p~value), max_size = bubble_max_size) +
      scale_fill_gradient2(low = "blue",
                           mid = "white",
                           high = "red") +
  hrbrthemes::theme_ipsum_es(base_family = myfont, base_size = basetextsize) +
  theme(axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        strip.background = element_blank(),
        legend.position = "bottom",
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  facet_wrap(~module_type, scales = "free_y", ncol = 2)
```

### Plot GSEA bubbleplots (Figures 4G of pre-print)

Red: enriched in protected/uninfected through 3 months post-vax surveillance
Blue: enriched in not protected/infected through 3 months post-vax surveillance

```{r plot gsea bubbleplot, fig.align='center', fig.width=7.333, fig.height=5, echo=TRUE, message=FALSE, warning=FALSE}
addSmallLegend <- function(myPlot, pointSize = 4, textSize = 6, spaceLegend = 0.6) {
    myPlot +
        guides(shape = guide_legend(override.aes = list(size = pointSize)),
               color = guide_legend(override.aes = list(size = pointSize))) +
        theme(strip.text = element_text(size=8),
              legend.direction = "vertical",
              legend.position = "right",
              legend.title = element_text(size = textSize), 
              legend.text  = element_text(size = textSize),
              legend.key.size = unit(spaceLegend, "lines"))
}
print(addSmallLegend(MyPlot))
```

```{r output to pdf, echo = FALSE, eval =FALSE}
# figdir <- "/Volumes/GoogleDrive/My Drive/Manuscripts/KSPZV1 Manuscript/Med Revised Submission 2022/Figures/Revised Figure Panels/"
# cairo_pdf(paste0(figdir, "Figure 4G voom version delta highdose only adj Batch Sex Site CSPAb_baseline MALdVAxTotal revised.pdf"), width = 6, height = 3.75)
# addSmallLegend(MyPlot)
# dev.off()
```

### Gene-level analysis

1. Identify genes that were differentially induced in uninfected and infected children receiving 1.8x10^6 PfSPZ Vaccine

2. Merge ΔP vs ΔNP, ΔP, and ΔNP tables.

3. Criteria:
    ΔP vs ΔNP P val < 0.005
    ΔP logFC > 0 and ΔP P val < 0.05
    ΔNP logFC < 0 and ΔNP P val < 0.10

```{r determine genes of interest, message=FALSE, warning=FALSE}
AllDegs1 <- bind_rows(DeltaDelta, .id = "Treatment") %>%
  mutate(Comparison = "DeltaDelta")
AllDegs2 <- bind_rows(DeltaProtected, .id = "Treatment") %>%
  mutate(Comparison = "DeltaProtected")
AllDegs3 <- bind_rows(DeltaNotProtected, .id = "Treatment") %>%
  mutate(Comparison = "DeltaNotProtected")
AllDegs <- bind_rows(AllDegs1, AllDegs2, AllDegs3) %>%
  dplyr::select(Treatment, Comparison, EnsemblID, GeneSymbol, descripton_new, logFC, AveExpr, "t", P.Value, adj.P.Val) %>%
  dplyr::rename(PValue = P.Value) %>%
  dplyr::rename(FDR = adj.P.Val) %>%
  dplyr::rename(Description = "descripton_new")

gois <- AllDegs %>%
  filter(Treatment == "1.8 x 10^6 PfSPZ") %>%
  droplevels() %>%
  dplyr::select(Treatment, Comparison, EnsemblID, GeneSymbol, Description, logFC, PValue, FDR) %>%
  pivot_wider(., names_from = Comparison, names_glue = "{Comparison}_{.value}", values_from = c(logFC, PValue, FDR)) %>%
  filter(DeltaDelta_PValue < 0.005 &
           (DeltaProtected_logFC > 0 & DeltaProtected_PValue < 0.05) &
           (DeltaNotProtected_logFC < 0 & DeltaNotProtected_PValue < 0.10)) %>%
  arrange(DeltaDelta_PValue)

gois_ms <- gois %>%
  dplyr::select(GeneSymbol, DeltaDelta_logFC, DeltaDelta_PValue, DeltaProtected_logFC, DeltaProtected_PValue, DeltaNotProtected_logFC, DeltaNotProtected_PValue)
```

### Figure 4H in pre-print

```{r table of goi, message=FALSE, warning=FALSE}
gois_ms %>%
  mutate_if(is.numeric, signif,3) %>%
  knitr::kable()
```

```{r output table of DEGS, echo=FALSE, eval=FALSE}
resdir <- "/Volumes/GoogleDrive/My Drive/Manuscripts/KSPZV1 Manuscript/Med Revised Submission 2022/Supplementary Tables/"
gois_ms %>%
  mutate_if(is.numeric, signif,3) %>%
  writexl::write_xlsx(paste0(resdir, "Table 4H Delta Post-vax DEGs Revised Adj Batch Sex Site log CSP Ab baseline MALdVaxTotal.xlsx"))
```

### Reshape and merge for survival analysis

```{r reshape and merge, echo = TRUE, message=FALSE, warning=FALSE}
library(gtools)
#from google drive
temp <- tempfile(fileext = ".rds")
dl <- drive_download(
  as_id("17xx0YaggLiyWdJc9rqA1nkPSE7dj05p0"), path = temp, overwrite = TRUE)
cpm_delta <- readRDS(file = dl$local_path)
cpm_delta <- cpm_delta[fData(cpm_delta)$GeneSymbol %in% gois_ms$GeneSymbol, ]
cpm_delta <- cpm_delta[,cpm_delta$treat != "Placebo"]
pData(cpm_delta) <- droplevels(pData(cpm_delta))
survdat <- Biobase::exprs(cpm_delta) %>%
  as.data.frame() %>%
  rownames_to_column(var = "EnsemblID") %>%
  left_join(., fData(cpm_delta) %>%
              dplyr::select(EnsemblID, GeneSymbol),
            by = "EnsemblID") %>%
  dplyr::select(-c(EnsemblID)) %>%
  dplyr::select(GeneSymbol, everything()) %>%
  pivot_longer(2:ncol(.), names_to = "PATID", values_to = "expression") %>%
  group_by(GeneSymbol) %>%
  mutate(exprs_up_down = factor(ifelse(expression>log2(1), "induced", "not induced"), levels = c("not induced", "induced"))) %>%
  mutate(exprs_terciles = factor(quantcut(expression, q=3, na.rm=TRUE), labels = c("bottom_tercile", "middle_tercile", "top_tercile"))) %>%
  mutate(exprs_median = factor(quantcut(expression, q=2, na.rm=TRUE), labels = c("bottom", "top"))) %>%
  left_join(., fData(cpm_delta) %>%
              dplyr::select(c(GeneSymbol, description)), by = "GeneSymbol") %>%
  left_join(., pData(cpm_delta)%>%
              dplyr::select(c(PATID, Timepoint, treat, site, SEX, pfcsp_pre, age.vax1, mal.vax.1, mal.dvax, mal.dvax.tot, contains("atp"))), by = "PATID") %>%
  droplevels()
survdat$tte.mal.atp.6 <- survdat$tte.mal.atp.6+1
survdat[is.na(survdat$pfcsp_pre),]$pfcsp_pre <- median(survdat$pfcsp_pre, na.rm = TRUE) #Impute 2 missing values with median of all samples
```

### Kaplan-Meier plots

FSTL4 is in Figure 4I of preprint

```{r km plots, echo=TRUE, message=FALSE, warning=FALSE}
library(survival)
library(survminer)
myFollowup <- 6 #3 or 6
N <- length(unique(survdat$PATID))
survdat.filtered <- fit <- mySurvPlot <- c()
for(i in gois_ms$GeneSymbol){
  survdat.filtered[[i]] <- survdat %>%
    filter(GeneSymbol == i)
  fit[[i]] <- survfit(Surv(tte.mal.atp.6, mal.atp.6) ~ exprs_up_down, data= survdat.filtered[[i]])
  mySurvPlot[[i]] <- ggsurvplot(fit[[i]], risk.table = TRUE, pval = TRUE,
                                break.time.by = 30, title = unique(survdat.filtered[[i]]$GeneSymbol),
                                xlab = "Days since vaccination",
                                ylab = "% free of parasitemia",
                                censor = TRUE,
                                palette = "lancet",
                                conf.int = TRUE,
                                conf.int.alpha = 0.1,
                                font.family = "Arial")
}
```


```{r print km plots, echo = FALSE, eval = FALSE, fig.align='center', fig.height = 5, fig.width = 6}
for(i in gois_ms$GeneSymbol){
  plot.new()
  print(mySurvPlot[[i]], newpage = FALSE)
}
```

```{r print invidividual plots, echo=FALSE, eval=FALSE}
# figdir <- "/Volumes/GoogleDrive/My Drive/Manuscripts/KSPZV1 Manuscript/Med Revised Submission 2022/Figures/"
# cairo_pdf(filename = paste0(figdir,"KSPZV1 Fig 4I Delta KM FSTL4 Revision.pdf"), width = 5.5, height = 5.5)
# print(mySurvPlot$FSTL4)
# dev.off()
```

### Fig 4I of preprint

```{r km plot 1, echo = FALSE, fig.align='center', fig.height=6, fig.width=6}
print(mySurvPlot$FSTL4, newpage = FALSE)
```

### Cox Regression

```{r do cox regression to separate dose groups, echo=TRUE, message=FALSE, warning=FALSE}
cox.res <- c()
for(i in names(survdat.filtered)){
  survdat.cp <- survdat.filtered[[i]]
  survdat.cp$tte.mal.atp.6 <- survdat.cp$tte.mal.atp.6 + 1 #add one day to each value given zero values
  survdat.cp$start		<-		0
  survdat.cp$stop		<-		survdat.cp$tte.mal.atp.6
  subj.pcrp <- rep(survdat.cp$start, as.vector(table(survdat.cp$PATID)))
  survdat.cp$start1 <- survdat.cp$start-subj.pcrp
  survdat.cp$stop1 <- survdat.cp$stop-subj.pcrp
  survdat.cp <- survdat.cp %>%
    droplevels()
  fit <- coxph(Surv(start1, stop1, mal.atp.6) ~ exprs_up_down + factor(SEX) + factor(site) + log10(pfcsp_pre+1) + mal.dvax.tot + factor(treat), data = survdat.cp)
  #vfit <- cox.zph(fit)
  #par(mfrow=c(3,2))
  #plot(vfit)
  #cox.zph(fit)
  cox.res[[i]] <- as.data.frame(cbind(fit$n, fit$nevent, summary(fit)$conf.int[,c(1,3:4)], summary(fit)$coefficient[,5]))
  colnames(cox.res[[i]]) <- c("n","number of events","HR", "LCI", "UCI", "P value")
  cox.res[[i]] <- cox.res[[i]] %>%
    as.matrix() %>%
    signif(., digits = 3) %>%
    as.data.frame()
  cox.res[[i]]$Gene <- i
  rownames(cox.res[[i]]) <- c("upregulated post-vax (ref:downregulated post-vax)",
                              "gender (ref:female)",
                              "Wagai (ref:Siaya)",
                              "CSP-specific IgG baseline",
                              "number of Pf infections during vaccination period",
                              "9.0x10^5 PfSPZ (ref:4.5x10^5 PfSPZ)",
                              "1.8x10^6 PfSPZ (ref:4.5x10^5 PfSPZ)")
  cox.res[[i]] <- cox.res[[i]] %>%
    rownames_to_column(var = "Covariate") %>%
    dplyr::select(Gene, everything())
  }
  cox.res.bound <- bind_rows(cox.res, .id = "Gene") %>%
    mutate(Significant = ifelse(`P value`<0.001, "***",
                              ifelse(`P value`<0.01, "**",
                                     ifelse(`P value`<0.05, "*", ""))))

cox.res.bound %>%
  dplyr::filter(`P value` < 0.05 & Covariate == "upregulated post-vax (ref:downregulated post-vax)") %>%
  arrange(HR,`P value`) 
```

### Table S4 of preprint

Only display FSTL4 here.

```{r show table, echo=TRUE}
cox.res.bound %>%
  filter(Gene == "FSTL4") %>%
  knitr::kable()
```


```{r write table to file, echo=FALSE, eval=FALSE}
resdir <- "/Volumes/GoogleDrive/My Drive/Manuscripts/KSPZV1 Manuscript/Med Revised Submission 2022/Supplementary Tables/"
genesbypvalue <- cox.res.bound %>%
  filter(Covariate=="upregulated post-vax (ref:downregulated post-vax)") %>%
  arrange(`P value`)

cox.res.bound %>%
  arrange(match(Gene, genesbypvalue$Gene)) %>%
  writexl::write_xlsx(., paste0(resdir, "Table S4 Cox PH for 22 genes n169 postvax 6 mos adj Batch Sex Site CSPAb MALdVAXTotal revised.xlsx"))
```

```{r save Rdata, echo=FALSE, eval=FALSE}
# save(xgroup, ygroup, design, fit, fit2, file = paste0("/Volumes/GoogleDrive/My Drive/Manuscripts/KSPZV1 Manuscript/Med Revised Submission 2022/Revision R Objects/", "KSPZV1-DGE-Delta-Postvax-voom-highdose-only-", Sys.Date(),".RData"))
```

